Document react query, profiling and react-scan (#1026)
software-mansion/radon-ide: packages/docs/docs/features/dev-tools.md

---
id: dev-tools
title: Dev Tools
sidebar_position: 7
---

Radon IDE comes with support for launching Dev Tools as separate panels withing VSCode or Cursor.

> The built-in tools for Network Inspector, Redux, Outline Rerenders (via [react-scan](https://react-scan.com/)) and React Query are integrated with the IDE and doesn't require any additional configuration.

Reviewer: ```suggestion
The built-in tools for Network Inspector, Redux, Render Outlines (via [react-scan](https://react-scan.com/)) and React Query are integrated with the IDE and doesn't require any additional configuration.
```
The tool is for outlining the renders, not rendering the outlines. 😉

> The built-in tools for Network Inspector, Redux, Outline Rerenders (via [react-scan](https://react-scan.com/)) and React Query are integrated with the IDE and doesn't require any additional configuration.

Reviewer 2: ```suggestion
The built-in tools for Network Inspector, Redux, Outline Renders (via [react-scan](https://react-scan.com/)) and React Query are integrated with the IDE and doesn't require any additional configuration.
```

> > NOTE: If your app is using [Expo Devtools Plugins](https://docs.expo.dev/debugging/devtools-plugins/) you will be able to use Redux and React Query dev tools via the plugin too. However, we recommend using the tools that are built into the Radon IDE instead of the one from the plugins as Radon relies on the official dev extension for Redux and React Query respecivelly.

Reviewer: ```suggestion
> NOTE: If your app uses [Expo Devtools Plugins](https://docs.expo.dev/debugging/devtools-plugins/) you will be able to use the Redux and React Query dev tools via the plugin too. However, we recommend using the tools that are built into the Radon IDE instead of the ones from the plugins as Radon relies on the official dev extension for Redux and React Query respectively.
```

<img width="700" src="/img/docs/ide_devtools_redux.png" className="shadow-image" />

> ## React Query

Reviewer 3: For react query and netowrk panel I wonder if we shouldn't mention that the events are captured **after** the panel is open.

This plugin doesn't require any additional configuration.
When the IDE detects your app uses React Query, the plugin will be listed in the Dev Tools menu.
Once enabled you can use the official [React Query's devtool UI](https://tanstack.com/query/v4/docs/framework/react/devtools) from within the editor panel:

<img width="700" src="/img/docs/ide_devtools_react_query.png" className="shadow-image" />

## CPU Profiling JavaScript

Radon IDE integrates with the hermes sampling CPU profiler.
When you app uses the supported version of hermes and devtools, the option to start profiling will appear in the Dev Tools menu.
Once profiling is started a button will appear next to the navigation bar in the IDE indicating the profiling is in progress.
You can use the button at the end of the profiling session to trigger a save dialog of the profiling output.
Once you save the profiling file, the IDE will automatically open the saved profile and use the built-in [Profile Visualizer](https://github.com/microsoft/vscode-js-profile-visualizer) plugin to display a profile.

<img width="550" src="/img/docs/ide_devtools_profiler_stop.png" className="shadow-image" />

> When profile is saved, Radon IDE resolves the source location of the methods captured during profiling session, such that you can navigate to code directly from the profile visualizer tool.

Reviewer: ```suggestion
When the profile is saved, Radon IDE resolves the source location of the methods captured during profiling session such that you can navigate to the code directly from the profile visualizer tool.
```

<img width="700" src="/img/docs/ide_devtools_profiler_visualize.png" className="shadow-image" />

> > NOTE: In order to view the profile in flame chart view, you need to install official [Flame Chart Visualizer](https://marketplace.visualstudio.com/items?itemName=ms-vscode.vscode-js-profile-flame) extension. Once the profile is saved, Radon IDE will display a message with direct installation link for this extension.

Reviewer 3: What about replacing all > NOTE: with :::info?

> When profiling, remember that the app you profile is the development build of your application running on a simulator environment.

Reviewer: ```suggestion
When profiling, remember that the app you profile is the development build of your application running in a simulator environment.
```

> When profiling, remember that the app you profile is the development build of your application running in a simulator environment.

Reviewer 3: Maybe :::danger?

> Development builds typically come with a significant overhead of extra work that is only done to provide better error reporting and tracability of errors.

Reviewer: ```suggestion
Development builds typically come with a significant overhead of extra work that is only done to provide better error reporting and traceability of errors.
```

> Hance the performance may significanly differ compared to the production builds running on an actual phone.

Reviewer: ```suggestion
Hence the performance may significantly differ compared to the production builds running on an actual phone.
```

You can enable this option by selecting "Outline Renders" from the Dev Tools menu.
Once enabled, the IDE will visualize React renders happening within the app.

> NOTE: Unlike with other dev tools, this setting turns off automatically when you close the editor or the Radon IDE panel.

<img width="700" src="/img/docs/ide_devtools_react_scan.png" className="shadow-image" />

## Redux (via Expo Devtools Plugin)

We recommend you use the built-in Redux dev tool as described in the above section which requires no extra setup.
However, if your workflow relies on using Expo CLI and you need to have the [Expo Redux Devtool Plugin](https://docs.expo.dev/debugging/devtools-plugins/#redux) setup in your project, Radon IDE will automatically detect it and will use the Redux devtools via the installed plugin.

<img width="700" src="/img/docs/ide_devtools_expo_redux.png" className="shadow-image" />

## React Query (via Expo Devtools Plugin)

We recommend you use the built-in React Query tool as described in one of the previous sections as it provides the official UI for React Query that is much more powerful than the one build for the Expo Plugin.

> For legacy reasons however, Radon IDE still supports the Expo Devtool Plugin for React Query.

Reviewer 3: I removed implementation for that, so I assume we should get rid of this section.

Or I can revert those chagnes.

Author: ok, I deleted that section and corresponding screenshots
